File: typescript/packages/host/src/ui/sse.ts
import { canonicalize } from "@threads/core/host";
import type { Protocol } from "./frame";
import type { Out } from "./stream";

// A UI stream on the wire: one SSE message per frame, its data the frame's chunk as canonical
// JSON (RFC 8785) and its id the frame's <seq>:<k> when it has one. An AI SDK stream that ends
// cleanly sends data: [DONE]; a broken one just closes, so the client falls back to a replay.

const HEADERS: Readonly<Record<Protocol, Readonly<Record<string, string>>>> = {
  "ai-sdk": {
    "content-type": "text/event-stream",
    "cache-control": "no-cache",
    "x-vercel-ai-ui-message-stream": "v1",
    "x-accel-buffering": "no",
  },
  "ag-ui": {
    "content-type": "text/event-stream",
    "cache-control": "no-cache",
    "x-accel-buffering": "no",
  },
};

export function sseResponse(
  protocol: Protocol,
  frames: AsyncGenerator<Out, void, undefined>,
): Response {
  const utf8 = new TextEncoder();
  const body = new ReadableStream<Uint8Array>({
    async pull(controller) {
      const next = await frames.next();
      if (next.done === true) {
        controller.close();
        return;
      }
      const out = next.value;
      if ("end" in out) {
        if (out.end === "done" && protocol === "ai-sdk")
          controller.enqueue(utf8.encode("data: [DONE]\n\n"));
        controller.close();
        await frames.return();
        return;
      }
      const data = canonicalize(out.data);
      if (!data.ok) throw new Error(`a frame is JSON: ${data.error.message}`);
      const id = out.id === undefined ? "" : `id: ${out.id}\n`;
      controller.enqueue(utf8.encode(`${id}data: ${data.value}\n\n`));
    },
    async cancel() {
      await frames.return();
    },
  });
  return new Response(body, { status: 200, headers: HEADERS[protocol] });
}

File: python/src/threads/host/ui/sse.py
"""A UI stream on the wire: one SSE message per frame, its data the frame's chunk as canonical
JSON (RFC 8785) and its id the frame's `<seq>:<k>` when it has one. An AI SDK stream that ends
cleanly sends `data: [DONE]`; a broken one just closes, so the client falls back to a replay."""

from collections.abc import AsyncGenerator, AsyncIterator, Mapping
from typing import Final

from threads.host.ui.frame import Protocol
from threads.host.ui.stream import End, Out
from threads.log.jcs import canonicalize
from threads.result import Ok

HEADERS: Final[Mapping[Protocol, Mapping[str, str]]] = {
    "ai-sdk": {
        "cache-control": "no-cache",
        "x-vercel-ai-ui-message-stream": "v1",
        "x-accel-buffering": "no",
    },
    "ag-ui": {"cache-control": "no-cache", "x-accel-buffering": "no"},
}


async def sse(protocol: Protocol, frames: AsyncGenerator[Out]) -> AsyncIterator[str]:
    try:
        async for out in frames:
            if isinstance(out, End):
                if out.how == "done" and protocol == "ai-sdk":
                    yield "data: [DONE]\n\n"
                return
            data = canonicalize(dict(out.data))
            if not isinstance(data, Ok):
                raise AssertionError(f"a frame is JSON: {data.error}")
            head = "" if out.id is None else f"id: {out.id}\n"
            yield f"{head}data: {data.value}\n\n"
    finally:
        await frames.aclose()
